src/sentry/models/authenticator.py

```diff
@@ -0,0 +1,472 @@
+    del _get_secret, _set_secret
+
+    def make_otp(self):
+        return TOTP(self.secret)
+
+    def validate_otp(self, otp):
+        otp = otp.strip().replace('-', '').replace(' ', '')
+        return self.make_otp().verify(otp)
+
+
+@register_authenticator
+class TotpInterface(OtpMixin, AuthenticatorInterface):
+    """This interface uses TOTP with an authenticator."""
+    type = 1
+    interface_id = 'totp'
+    name = _('Authenticator App')
+    description = _('An authenticator application that supports TOTP (like '
+                    'Google Authenticator or 1Password) can be used to '
+                    'conveniently secure your account.  A new token is '
+                    'generated every 30 seconds.')
+
+    def get_provision_qrcode(self, user, issuer=None):
+        return self.make_otp().get_provision_qrcode(
+            user, issuer=issuer)
+
+
+@register_authenticator
+class SmsInterface(OtpMixin, AuthenticatorInterface):
+    """This interface sends OTP codes via text messages to the user."""
+    type = 2
+    interface_id = 'sms'
+    name = _('Text Message')
+    description = _('This authenticator sends you text messages for '
+                    'verification.  It\'s useful as a backup method '
+                    'or when you do not have a phone that supports '
+                    'an authenticator application.')
+    code_ttl = 45
+
+    @property
+    def is_available(self):
+        return sms_available()
+
+    def generate_new_config(self):
+        config = super(SmsInterface, self).generate_new_config()
+        config['phone_number'] = None
+        return config
+
+    def make_otp(self):
+        return TOTP(self.config['secret'], digits=6, interval=self.code_ttl,
+                    default_window=1)
+
+    def _get_phone_number(self):
+        return self.config['phone_number']
+
+    def _set_phone_number(self, value):
+        self.config['phone_number'] = value
+
+    phone_number = property(_get_phone_number, _set_phone_number)
+    del _get_phone_number, _set_phone_number
+
+    def activate(self, request):
+        if self.send_text(request=request):
+            return ActivationMessageResult(
+                _('A confirmation code was sent to your phone. '
+                  'It is valid for %d seconds.') % self.code_ttl)
+        return ActivationMessageResult(
+            _('Error: we failed to send a text message to you. You '
+              'can try again later or sign in with a different method.'),
+            type='error')
+
+    def send_text(self, for_enrollment=False, request=None):
+        ctx = {'code': self.make_otp().generate_otp()}
+
+        if for_enrollment:
+            text = _('You are about to set up two-factor authentication '
+                     'through text messages. Your confirmation code is '
+                     '%(code)s.')
+        else:
+            text = _('Your Sentry authentication code is %(code)s.')
```

Comment:
We should change this around to put the code first. For example, GitHub says "%(code)s is your GitHub authentication code."

This is important since you want to make sure the code shows up in a notification bar or whatever first so it's easier to use.

Google does this as well.